File: spring/web/SpringMvcTesting/src/main/java/com/instil/mvc/logic/QuotesEngineImpl.java
package com.instil.mvc.logic;

import com.instil.mvc.logic.QuotesEngine;
import com.instil.mvc.model.Course;
import com.instil.mvc.model.Quote;
import org.springframework.stereotype.Service;

import javax.annotation.Resource;
import java.util.Map;
import java.util.Optional;

@Service
public class QuotesEngineImpl implements QuotesEngine {

    @Override
    public Optional<Quote> generateQuote(String courseId) {
        return findCourse(courseId)
                .map(this::buildQuote);
    }

    private Quote buildQuote(Course course) {
        double rate = 0;
        switch(course.getDifficulty()) {
            case BEGINNER:
                rate = 1200.0;
                break;
            case INTERMEDIATE:
                rate = 1400.0;
                break;
            case ADVANCED:
                rate = 1600.0;
                break;
        }
        return new Quote(course, course.getDuration() * rate);
    }

    private Optional<Course> findCourse(String courseId) {
        return portfolio.values()
                .stream()
                .filter(c -> c.getId().equals(courseId))
                .findFirst();
    }

    @Resource(name = "portfolio")
    public void setPortfolio(Map<String, Course> portfolio) {
        this.portfolio = portfolio;
    }

    private Map<String, Course> portfolio;
}

File: spring/web/SpringMvcTestingKotlin/src/main/kotlin/demos/spring/boot/rest/logic/QuotesEngineImpl.kt
package demos.spring.boot.rest.logic

import demos.spring.boot.rest.model.Course
import demos.spring.boot.rest.model.CourseDifficulty.*
import demos.spring.boot.rest.model.Quote
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service

@Service
class QuotesEngineImpl(@Qualifier("portfolio") val portfolio: MutableMap<String, Course>) : QuotesEngine {
    override fun generateQuote(courseId: String): Quote? {
        val course = portfolio[courseId]
        return if(course != null) buildQuote(course) else null
    }

    private fun buildQuote(course: Course): Quote {
        val rate = when (course.difficulty) {
            BEGINNER -> 1200.0
            INTERMEDIATE -> 1400.0
            ADVANCED -> 1600.0
        }
        return Quote(course, course.duration * rate)
    }

}
